gunpowder/jax/nodes/train.py

```diff
@@ -0,0 +1,288 @@
+            self.model_params, outputs, loss = jax.pmap(
+                                            self.model.train_step,
+                                            axis_name='num_devices',
+                                            donate_argnums=(0,),
+                                            static_broadcasted_argnums=(2,))(
+                                    self.model_params, inputs, True)
+            loss = loss.mean()
+            outputs = self.unstack_device_outputs(outputs)  # stack by batch
+        else:
+            self.model_params, outputs, loss = jax.jit(
+                                            self.model.train_step,
+                                            donate_argnums=(0,),
+                                            static_argnums=(2,))(
+                                    self.model_params, inputs, False)
+
+        logger.debug(
+            "model outputs: %s",
+            {k: v.shape for k, v in outputs.items()})
+
+        # add requested model outputs to batch
+        for array_key, array_name in requested_outputs.items():
+            spec = self.spec[array_key].copy()
+            spec.roi = request[array_key].roi
+            batch.arrays[array_key] = Array(
+                outputs[array_name], spec
+            )
+
+        batch.loss = loss
+        self.iteration += 1
+        batch.iteration = self.iteration
+
+        if batch.iteration % self.save_every == 0:
+
+            checkpoint_name = self._checkpoint_name(
+                self.checkpoint_basename, batch.iteration)
+
+            logger.info("Creating checkpoint %s", checkpoint_name)
+
+            model_state = self.model_params
+            if self.n_devices > 1:
+                # get only a single copy of param for saving
+                model_state = jax.tree_map(lambda x: x[0], model_state)
+            with open(checkpoint_name, 'wb') as f:
+                pickle.dump(model_state, f)
+
+            if self.keep_n_checkpoints:
+                checkpoint_name = self._checkpoint_name(
+                    self.checkpoint_basename,
+                    batch.iteration - self.keep_n_checkpoints*self.save_every)
+                try:
+                    os.remove(checkpoint_name)
+                    logger.info("Removed checkpoint %s", checkpoint_name)
+                except FileNotFoundError:
+                    pass
+
+        if self.summary_writer and batch.iteration % self.log_every == 0:
+            self.summary_writer.add_scalar("loss", batch.loss, batch.iteration)
+
+    def __collect_requested_outputs(self, request):
+
+        array_outputs = {}
+
+        for output_name, array_key in self.outputs.items():
+            if array_key in request:
+                array_outputs[array_key] = output_name
+
+        return array_outputs
+
+    def __collect_provided_inputs(self, batch):
+
+        return self.__collect_provided_arrays(
+            {k: v for k, v in self.inputs.items()}, batch
```

Comment:
Whats this dict comprehension for?

Reply:
I think this is for getting items from a batch to a `Dict`. I just basically copied this from the Torch train node: https://github.com/funkey/gunpowder/blob/master/gunpowder/torch/nodes/train.py#L332

Reply:
Looking at it I guess it's redundant :sweat_smile: . Will fix it.

Reply:
:+1: I may have written it in the first place if it came from the torch node. Probably better not to propagate it